### ISSUE TITLE
Conflict with multiple CIntercomService?

### ISSUE THREAD

**Issue by rbx** [FairRootGroup/DDS#154]
In my application I am using both `CKeyValue` and `CCustomCmd`, that are both loaded as plugins. Inside of the plugin they both live as a member of a singleton. The singleton of both plugins contains `CIntercomService` as a member.

Each plugin exposes 3 functions: init, run, stop (actual names are different):
- init() subscribes for `CKeyValue`/`CCustomCmd` (and also creates the instance).
- run() runs `CIntercomService.start()` and `CKeyValue.PutValue()`.
- stop() deletes the instance to make sure destructors of its members are called. 

Before my application quits, it calls stop() of each plugin followed by `dlclose()` on each plugin library.

I am facing two issues when doing the above:
1. When running on MacOS (10.12.1 + Apple LLVM 8.0.0), after both plugins are loaded the application constantly consumes 100% CPU (plugins are working properly!). This does not happen on a Linux machine, where CPU consumption is normal and plugins also do their job.
2. When plugins are trying to close, the deletion of the singleton (destructor of the service called) of the first plugin hangs indefinitely on MacOS. On Linux the plugins sometimes close, sometimes not. Regardless of whether they closed or not, they sometimes also cause following error:

```
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::thread_resource_error> >'
  what():  boost thread: trying joining itself: Resource deadlock avoided
```

None of the above problems occur if I load only one plugin instead of two. In this case everything is working correctly.

**Comment by AnarManafov** [comment]
Alexey, can you please try to compile your main executables using "-rdynamic" linker option?
Let me know whether it helps or not.

**Comment by rbx** [comment]
I tried to compile my executables and plugins with the following:
```
cmake -DCMAKE_EXE_LINKER_FLAGS="-rdynamic" -DCMAKE_SHARED_LINKER_FLAGS="-rdynamic" ..
```

The picture remained the same. Or maybe I pass the flag wrong?

**Comment by rbx** [comment]
To narrow down the cause I tried to reproduce the issue with significantly simplified executables and plugins.

I start this topology:
https://github.com/rbx/FairRoot/blob/ddstest/examples/MQ/3-dds/ex3-dds-topology.xml

With two executables, both sharing the same code:
https://github.com/rbx/FairRoot/blob/ddstest/examples/MQ/3-dds/runExample3Sampler.cxx
https://github.com/rbx/FairRoot/blob/ddstest/examples/MQ/3-dds/runExample3Sink.cxx

And the simplified config (CKeyValue)/control (CCustomCmd) plugin code:
https://github.com/rbx/FairRoot/blob/ddstest/fairmq/plugins/config/FairMQDDSConfigPlugin.cxx
https://github.com/rbx/FairRoot/blob/ddstest/fairmq/plugins/control/FairMQDDSControlPlugin.cxx

When activating this topology I see the same issues as described in the original report (linux and MacOS). On MacOS the closing ResetInstance() function of the plugin hangs and both processes consume 100% CPU.

I also try to run one of the above executable by itself, without having started the server. Now there is no 100% CPU usage on Mac and the shutdown process seems to proceed much further:
```
[21:35:55][DEBUG] Opening config plugin: libFairMQDDSConfigPlugin.dylib
[21:35:55][DEBUG] Opening control plugin: libFairMQDDSControlPlugin.dylib
[21:35:55][DEBUG] [FairMQConfigPluginDDS] Subscribing for DDS properties.
[21:35:55][DEBUG] [FairMQControlPluginDDS] Subscribing for commands.
[21:35:55][INFO] initialized plugins
[21:35:55][ERROR] [FairMQConfigPluginDDS] DDS key-value error code: 1, message: Error in the transport service: Cannot find DDS commander info file.
[21:35:55][ERROR] [FairMQControlPluginDDS] Error received: error code: 1, error message: Error in the transport service: Cannot find DDS commander info file.

[21:35:55][DEBUG] [FairMQConfigPluginDDS] Publishing 'bla' to DDS under 'data' property name.
[21:35:55][ERROR] [FairMQConfigPluginDDS] DDS key-value error code: 3, message: CCDDSIntercomGuard: Shared memory channel is not running. Failed to put the key-value <data --> bla
[21:35:55][ERROR] [FairMQControlPluginDDS] Error received: error code: 3, error message: CCDDSIntercomGuard: Shared memory channel is not running. Failed to put the key-value <data --> bla

[21:35:55][INFO] [FairMQControlPluginDDS] Listening for commands from DDS...
[21:35:55][INFO] running ... 0
...
[21:36:09][INFO] running ... 14
[21:36:10][DEBUG] Closing FairMQConfigPlugin...
[21:36:10][DEBUG] [FairMQConfigPluginDDS]: Resetting instance.
[21:36:10][DEBUG] [FairMQConfigPluginDDS] Unsubscribing.
[21:36:10][DEBUG] [FairMQConfigPluginDDS]: Instance has been reset.
[21:36:10][DEBUG] Closing FairMQControlPlugin...
[21:36:10][DEBUG] [FairMQControlPluginDDS]: Resetting instance.
[21:36:10][DEBUG] [FairMQControlPluginDDS] Unsubscribing
[21:36:10][DEBUG] [FairMQControlPluginDDS]: Instance has been reset.
Segmentation fault: 11
```

If I start the server but do not submit, the picture is same as the previous, but with slightly fewer ERRORs:
```
...
[21:22:37][DEBUG] [FairMQConfigPluginDDS] Publishing 'bla' to DDS under 'data' property name.
[21:22:37][ERROR] [FairMQConfigPluginDDS] DDS key-value error code: 3, message: CCDDSIntercomGuard: Shared memory channel is not running. Failed to put the key-value <data --> bla
[21:22:37][ERROR] [FairMQControlPluginDDS] Error received: error code: 3, error message: CCDDSIntercomGuard: Shared memory channel is not running. Failed to put the key-value <data --> bla
...
```

With an added submit, but no activate, output is same as previous one.

So it looks like running outside the proper environment or without access to the shared memory channel hides the problem (both CPU usage and shutdown). Although I don't see where the final `segmentation fault: 11` comes from.

You can run my executables outside of FairRoot to reproduce. I only depend on FairMQLogger, which can be easily replaced with `cout`.

**Comment by AnarManafov** [comment]
This issue should be fixed starting from DDS v3.0.
CIntercomService doesn't use a singleton anymore.
